lib/ansible/modules/clustering/kubevirt/k8s_service.py

```diff
@@ -0,0 +1,203 @@
+            - Use to specify a Service object namespace.
+        required: true
+        type: str
+    type:
+        description:
+            - Specifies the type of Service to create.
+            - https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types
+        choices:
+            - NodePort
+            - ClusterIP
+            - LoadBalancer
+            - ExternalName
+        required: true
+    ports:
+        description:
+            - A list of ports to expose.
+            - https://kubernetes.io/docs/concepts/services-networking/service/#multi-port-services
+        required: true
+        type: list
+    selector:
+        description:
+            - Label selectors identify objects this Service should apply to.
+            - https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
+        required: true
+        type: dict
+
+extends_documentation_fragment:
+  - k8s_auth_options
+  - k8s_resource_options
+
+requirements:
+  - python >= 2.7
+  - openshift >= 0.6.2
+'''
+
+EXAMPLES = '''
+- name: Expose https port with ClusterIP
+  k8s_service:
+    state: present
+    name: test-https
+    namespace: default
+    type: ClusterIP
+    ports:
+    - port: 443
+      protocol: TCP
+    selector:
+      key: special
+'''
+
+RETURN = '''
+result:
+  description:
+  - The created, patched, or otherwise present Service object. Will be empty in the case of a deletion.
+  returned: success
+  type: complex
+  contains:
+     api_version:
+       description: The versioned schema of this representation of an object.
+       returned: success
+       type: str
+     kind:
+       description: Always 'Service'.
+       returned: success
+       type: str
+     metadata:
+       description: Standard object metadata. Includes name, namespace, annotations, labels, etc.
+       returned: success
+       type: complex
+     spec:
+       description: Specific attributes of the object. Will vary based on the I(api_version) and I(kind).
+       returned: success
+       type: complex
+     status:
+       description: Current status details for the object.
+       returned: success
+       type: complex
+     items:
+       description: Returned only when multiple yaml documents are passed to src or resource_definition
+       returned: when resource_definition or src contains list of objects
+       type: list
+'''
+
+import copy
+import traceback
+
+from collections import defaultdict
+
+from ansible.module_utils.k8s.common import AUTH_ARG_SPEC, COMMON_ARG_SPEC
+from ansible.module_utils.k8s.raw import KubernetesRawModule
+
+
+SERVICE_ARG_SPEC = {
+    'merge_type': {'type': 'list', 'choices': ['json', 'merge', 'strategic-merge']},
+    'selector': {'type': 'dict', 'default': {}},
+    'type': {
+        'type': 'str',
+        'choices': [
+            'NodePort', 'ClusterIP', 'LoadBalancer', 'ExternalName'
+        ],
+    },
+    'ports': {'type': 'list', 'default': []},
```

Comment:
Remeber that those default will always reset already existing services.

Reply:
Just to make sure we're understanding each other. If I have a config like this:

```
type: ClusterIP
ports:
  - port: 22
selector:
  key: special
```

And for whatever reason want to keep the clusterip, but stop exposing any ports (doesn't make sense in this case, but we might be doing other modules where it will):

```
type: ClusterIP
selector:
  key: special
```

then yes, I'd expect the Service in the cluster to match 1:1 the playbook, which in this case means: delete the ports if they're no longer in the playbook.

Is that not correct behavior?

Reply:
Yes this will do exactly what you described, but what if you want to update selector, and you don't want to specify all the current details of the service, but just want to define selector and name of the service, because it's easier. You will always have to define ports, otherwise you will change the setup.

Reply:
I went through k8s docs and k8s ansible module code and you are most likely correct that not enforcing this the way I did is the right way to do it. I'll push the changes tomorrow morning.

Reply:
Ok, pushed the patch without those defaults.